Sources/App/Controllers/API/API+PackageController+ForkedFromResult.swift

```diff
@@ -0,0 +1,42 @@
+// Copyright Dave Verwer, Sven A. Schmidt, and other contributors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import Fluent
+import SQLKit
+import Vapor
+
+extension API.PackageController {
+     enum ForkedFromResult {
```

Comment:
`ForkedFromResult` and `ForkedFromInfo` seem very similar. They have the same cases, almost the same associated values - could they be the same type? Is there a reason they're not? If so, perhaps the name could/should carry that concept.

Reply:
They are very similar. The reason I have two types is because I thought that it might be a good thing to separate the actual result and then to initialize the model (that's gonna be used in the view layer). Same thing if you look at `PackageResult` and `API.PackageController.GetRoute.Model`. I'm not really sure. Should I just merge it into one type?

Reply:
`PackageResult` (a db model reference type) and `API.PackageController.GetRoute.Model` (a view model value type) are great examples of why and when to add these types.

We could have avoided transforming data into the various types that make up `GetRoute.Model` but we wanted to have a plain value type view model that we can use to drive a page. That's a clearer separation but it also allows us to very easily assemble data for snapshot testing etc.

(It was also really helpful early on when Dave designed the pages and just specified in a struct what data he needed and I populated it from the db results independently. It made for a great interface.)

You'll see that some types don't get translated that way. For example we pass `App.Reference` and others 1:1 to `GetRoute.Model` - and that's because they're plain value types already. They're easy to construct for the tests and equally simple to pass around.

`ForkedFromXXX` seems to be the same: it's a value type and it really seems to be packaging all the same data. So it can just be the same type - we wouldn't need to transform it unless there's some other reason for it.

One such reason might be some computation that's happening. I can't think of a great example right now but we'd want to avoid having `GetRoute.Model` do a lot other than just package up data for a page.

Does that make sense?

I can take a closer look myself to see if the types can be collapsed if you prefer, just let me know.

Reply:
The only difference `FrokedFromInfo` has from `ForkedFromResult` is this

```swift
      var url: String {
            switch self {
            case .fromSPI(_, let originalOwner, _, let originalRepo, _):
                return SiteURL.package(.value(originalOwner), .value(originalRepo), nil).relativeURL()
            case .fromGitHub(let url):
                return url
            }
        }

        var ownerURL: String? {
            switch self {
            case .fromSPI(_, let owner, _, _, _):
                return SiteURL.author(.value(owner)).relativeURL()
            case .fromGitHub:
                return nil
            }
        }
```

it has some computed variables which come in handy in the view itself - so I think they are totally mergeable. But where should the  resulting `ForkedFrom` be declared? And would be it be okay to make the `fetchForkedFromResult` method inside `API+PackageController+GetRoute` return the new `ForkedFrom` type and then pass it down to the Model through the initializer.

Could you please take a closer look when you find the time and propose the change, since I am not entirely sure what's the correct approach.

Reply:
I've opened a PR on top of your branch with the changes here: https://github.com/SwiftPackageIndex/SwiftPackageIndex-Server/pull/3388

That should sort it all out!

Reply:
Thanks a lot! The comments on the PR were really helpful! I've merged it into the branch.